### PR TITLE
Build LABjs string from source

### PR DESCRIPTION
**Pull request by gigabo**
I'm going to start making substantive changes to the LABjs code we put at the top of every page.  I want to have useful diffs and commit history for those changes, so as a baseline this PR replaces the pre-minified source string with the (mostly) vanilla [source file](https://github.com/getify/LABjs/blob/master/LAB.src.js).

A few benefits:

- Can opt-in to viewing debug output with `DEBUG_LAB=true react-server start`
- It's 210 bytes smaller (I'm going to make it smaller still in follow up PRs)
- It's human-readable (and hackable)

A few drawbacks:

- A dep on uglify
- A few hundred millisecond startup penalty

Could minify during build to avoid those, but we'd lose the debug-ability which is really nice.

### CODE REVIEW

**Comment by gigabo** [comment]
I guess we could pre-build _two_ copies (normal and debug) and select among them with an option... 🤔 

That would actually let us turn on debugging at run-time which could be useful.

**Comment by gigabo** [comment]
Heh. 👹 

image

**Comment by doug-wade** [comment]
:thinking: iiuc, we're just checking in the source of lab.js to react server core, but is that really where it belongs?  Maybe it makes sense to make it a separate npm package?  Maybe it even makes sense to make it into an npm package that lives outside of the react server repo, maybe something like https://npmjs.com/package/labjs?  I'm not sure if the license would permit that, but it doesn't really seem like its something react server specific...

**Comment by doug-wade** [comment]
Actually, looks like there's [plans to put it on npm](https://github.com/getify/LABjs/issues/113#issuecomment-236385245)

**Comment by gigabo** [comment]
Oh, good thought.  I'm going to make some modifications for its use in React Server, so if anything it would probably have to be `react-server-labjs` or something like that.  But really, I'm not sure that it would be useful externally?  It's basically part of the React Server runtime in the browser.

**Comment by doug-wade** [comment]
Makes sense.  I vote we pry it out into `react-server-labjs` (which also means you can postpone fixing the linting errors) and hope that we can take a dependency on the npm package when it comes out.

**Comment by gigabo** [comment]
> hope that we can take a dependency on the npm package when it comes out.

We already can't.  We would need https://github.com/getify/LABjs/pull/81 to land for non-debug.

**Comment by gigabo** [comment]
But anyway once _this_ (or whatever this turns into) lands I've got another PR coming on its heels that starts _really_ hacking it up. 😈

File: packages/react-server/core/util/LABString.js [hunk 1]
```diff
@@ -1,11 +1,26 @@
-// Careful: We've got a local patch here!
-//
-// We've added support for a `crossOrigin` parameter.
-//
+const path = require("path");
+const {readFileSync} = require("fs");
+
+var lab = readFileSync(path.join(__dirname, "LAB.src.js"), "utf-8");
```
Comment (drewpc) [inline packages/react-server/core/util/LABString.js]: Doesn't this mean that this module can't be used on the client side?

File: packages/react-server/core/util/LABString.js [hunk 2]
```diff
@@ -1,11 +1,26 @@
...
+
+// Run with `DEBUG_LAB=true` to see its output in console.log.
+if (!process.env.DEBUG_LAB) { // eslint-disable-line
+
+	// This is exactly what @getify does to remove debug code.
+	// https://github.com/getify/LABjs/blob/master/BUILD.md
+	lab = lab.replace(/\/\*!START_DEBUG(?:.|[\n\r])*?END_DEBUG\*\//g, "");
+
+	// After the debug code is removed just uglify.
+	lab = require("uglify-js").minify(lab, {fromString: true}).code;
```
Comment (drewpc) [inline packages/react-server/core/util/LABString.js]: We already run this whole package through Babel with Gulp, why not uglify it there and save the uglified version in `target/(client|server)/util/LABString.js`?  What am I missing?

File: packages/react-server/core/util/LABString.js (hunk 1 above)
Comment (gigabo) [inline packages/react-server/core/util/LABString.js]: Yep, this module is only used in the render middleware.  Its output is a string of JS that is put into an inline script tag.

File: packages/react-server/core/util/LABString.js (hunk 2 above)
Comment (gigabo) [inline packages/react-server/core/util/LABString.js]: > What am I missing?

The full extent of my laziness. 👹 

As I think about it more I like @doug-wade's idea of pulling this out into a separate package.  Going to manage it inside the React Server repo for now just for convenience.  We'll be able to pull it out later using [`lerna export`](https://github.com/lerna/lerna/pull/486) if we want to.

**Comment by gigabo** [comment]
I've updated this PR to put LAB into a dedicated package and move the minification to build time. 🍻

**Comment by gigabo** [comment]
Hard to believe [`flab`](https://www.npmjs.com/package/flab) wasn't taken. 😈

File: packages/flab/stringify.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+console.log(
+	'module.exports = '+
+	require("fs").readFileSync("/dev/stdin", "utf-8")
```
Comment (roblg) [inline packages/flab/stringify.js]: I feel like I should know this, but... do we officially support windows? Seems like `/dev/stdin` wouldn't work there?

Comment (gigabo) [inline packages/flab/stringify.js]: Ugh... this is only at build time.  Do we need to support windows even for _that_? 😩

Comment (drewpc) [inline packages/flab/stringify.js]: Great catch, @roblg.  @gigabo use the `memory-stream` package for that!  Examples:

- mock socket: https://github.com/redfin/react-server/blob/master/packages/react-server/core/__tests__/reactMiddlewareSpec.js
- mock fs: https://github.com/redfin/react-server/blob/master/packages/react-server-cli/src/__tests__/compileClient/compileClientSpec.js

Comment (drewpc) [inline packages/flab/stringify.js]: Ignore the `mockery` stuff.  The `memory-stream` package allows you to pass a stream object to a library and then call `.toString()` on what was written to the stream.

Comment (gigabo) [inline packages/flab/stringify.js]: Oh cool thanks @drewpc.

I'm tempted to use [`get-stdin`](https://www.npmjs.com/package/get-stdin), which is a bit more targeted to what this little one-liner needs to do.

**Comment by gigabo** [comment]
Okay, got rid of the use of `/dev/stdin`. 👍

**Comment by drewpc** [comment]
While you're involved in this, it probably is a good idea to look into https://github.com/getify/LABjs/issues/113.  It might address #77 and #253 without much work from this end.

**Comment by gigabo** [comment]
Oh, nice.  Thanks for the link.

File: packages/flab/README.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+Future Loading and Blocking JS
+
+This is based on the awesome [LABjs](https://github.com/getify/LABjs) package.
+
+TODO: Flesh out this README. :grin:
```
Comment (doug-wade) [inline packages/flab/README.md]: 😆

File: packages/flab/package.json [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{
+  "name": "flab",
```
Comment (doug-wade) [inline packages/flab/package.json]: 😆

File: packages/flab/package.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  },
+  "keywords": [
+    "loading",
+    "blocking",
+    "js",
+    "asynchronous",
+    "loader"
+  ],
+  "author": "Bo Borgerson",
+  "license": "MIT",
```
Comment (doug-wade) [inline packages/flab/package.json]: Apache 2?

File: packages/flab/package.json [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  "bugs": {
+    "url": "https://github.com/redfin/react-server/issues"
+  },
+  "homepage": "https://github.com/redfin/react-server#readme",
```
Comment (doug-wade) [inline packages/flab/package.json]: out-of-scope: should we update all of the modules to point to react-server.io as their homepage?

File: packages/flab/package.json (hunk 6 above)
Comment (gigabo) [inline packages/flab/package.json]: LABjs is MIT.  Can MIT be converted to Apache 2?

File: packages/flab/package.json (hunk 7 above)
Comment (gigabo) [inline packages/flab/package.json]: Oh, maybe... 🤔

**Comment by gigabo** [comment]
Onward! 👉

**Pull request merged**